### DISCUSSION TITLE
Background job information

### DISCUSSION THREAD

**Discussion by HasEveryUsernameBeenTaken** [JanDeDobbeleer/oh-my-posh#1495] [Ideas]
Would it be possible to display the number of active background jobs in a segment or include this in an existing segment?

I think I first saw this kind of thing in Liquid Prompt on GNU/Linux.  In PowerShell, I previously included the following in a custom prompt function, and I miss it in Oh My Posh.

```
[int] $BackgroundJobs = (Get-Job -State Running).Count
[int] $StoppedJobs = (Get-Job | Where-Object { 'Stopped','Stopping' -iContains $PSItem.State}).Count

…

If ($BackgroundJobs -gt 0) {
	Write-Host -NoNewline "${BackgroundJobs}& "
}
If ($StoppedJobs -gt 0) {
	Write-Host -NoNewline "${StoppedJobs}z "
}
```

**Comment by JanDeDobbeleer** [comment]
You can add this by leveraging the [text segment](https://ohmyposh.dev/docs/text). You can create a function to store the value in an environment variable:

```powershell
function Set-EnvVar {
    $env:BG_JOBS = (Get-Job -State Running).Count
}
New-Alias -Name 'Set-PoshContext' -Value 'Set-EnvVar' -Scope Global -Force
```

then add the following to your config:

```json
{
  "type": "text",
  "style": "powerline",
  "powerline_symbol": "\uE0B0",
  "foreground": "#ffffff",
  "background": "#0077c2",
  "properties": {
    "template": "{{if ne .Env.BG_JOBS \"0\" }}{{ .Env.BG_JOBS }}{{ end }}"
  }
}
```

**Reply by MattKotsenas** [reply]
Would you be willing to accept a contribution to make this a built-in segment? I'm mostly a PowerShell user, but it seems almost all supported shells have a similar background job functionality, so it feels like this segment would be broadly applicable.

**Reply by JanDeDobbeleer** [reply]
@MattKotsenas depends how long this cmdlet takes to execute. Otherwise it's a bad idea to add that in the core as this can't be measured with `oh-my-posh debug`.

**Reply by MattKotsenas** [reply]
`Get-Job` appears to be reasonably fast (at least for me; Surface Book 3 from 2020 @ 2.0 GHz):

```powershell
@(1..10) | %{ Measure-Command -Expression { Get-Job } } | %{ $_.TotalMilliseconds } | measure-object -Average

Count             : 10
Average           : 0.26698
Sum               :
Maximum           :
Minimum           :
StandardDeviation :
Property          :
```

A more realistic measure of the full end-to-end cost:

### With 0 jobs

```powershell
@(1..10) | %{ measure-command { $jobs = (Get-Job); $running = ($jobs | ?{ $_.State -eq "Running" }); $finished = ($jobs.Count - $running.Count); } } | %{ $_.TotalMilliseconds } | measure-object -average

Count             : 10
Average           : 1.91483
Sum               :
Maximum           :
Minimum           :
StandardDeviation :
Property          :
```

### With 500 jobs

```powershell
@(1..10) | %{ measure-command { $jobs = (Get-Job); $running = ($jobs | ?{ $_.State -eq "Running" }); $finished = ($jobs.Count - $running.Count); } } | %{ $_.TotalMilliseconds } | measure-object -average

Count             : 10
Average           : 15.13528
Sum               :
Maximum           :
Minimum           :
StandardDeviation :
Property          :
```

Bash shows a similar time:

```bash
start=`date +%s%N`; jobs -l ; end=`date +%s%N`; echo Execution time was $((($end - $start)/1000000)) milliseconds.
Execution time was 4 milliseconds.
```


For some context / comparison, on this machine if I get the value of `prompt`, comment out the call to `Set-PoshContext` as well as `oh-my-posh` itself, I get 4 - 5 milliseconds from `Measure-Command`.

That feels to me like a small enough cost to be tolerable (assuming sufficient interest). What else should I be considering?

Also, since I'm new to omp internals, is the rationale for not being able to measure it with `debug` because in order to get the user's session info, it would need to be part of `prompt` and not inside the omp Go code? Or in other words, users would pay for the jobs call even if they weren't using the segment?

**Reply by JanDeDobbeleer** [reply]
Decided not to do this as it does add load time on other shells.

**Reply by MattKotsenas** [reply]
Ok, fair enough. Thanks for updating the thread & letting me know!

**Reply by JanDeDobbeleer** [reply]
@MattKotsenas on second thought. I'm going to see if I can add it but toggle only when used.

**Comment by akelge** [comment]
A note about zsh (as I just fought to have it working)
```
  - type: text
    style: plain
    foreground: green
    properties:
      prefix: ""
      postfix: ""
      text: "$([ $(jobs|wc -l|xargs) -gt 0 ] && echo '\u2622 ')"
 ```
Here is how to do it in zsh, you can't use `command` segment, because it is executed in a subshell that does not know about BG jobs, but the above is executed inline and works

**Reply by JanDeDobbeleer** [reply]
You can reuse the same logic as indicated above using `set_poshcontext` in zsh:

```
function set_poshcontext() {
    export BG_JOBS=$(jobs|wc -l|xargs)
}
```
